src/main/java/com/google/sps/data/Room.java

```diff
@@ -0,0 +1,150 @@
+    private Queue<Video> videos;
+
+    /**
+      * Room constructor
+      * @param members a List of Member objects
+      * @param videos an Queue of video objects
+      * @return a new Room object
+      */
+    private Room(List<Member> members, Queue<Video> videos){
+        this.messages = new LinkedList<Message>();
+        this.members = members;
+        this.videos = videos;
+    }
+
+    /**
+      * Room constructor
+      * @param members a List of Member objects
+      * @param videos an Queue of video objects
+      * @param messages a List of message objects
+      * @return a new Room object
+      */
+    private Room(List<Member> members, Queue<Video> videos, LinkedList<Message> messages) {
+        this.members = members;
+        this.videos = videos;
+        this.messages = messages;
+    }
+
+    //Room factory function
+    public static Room createRoom(List<Member> members, Queue<Video> videos, LinkedList<Message> messages){
+        return new Room(members, videos, messages);
+    }
+
+    //Returns the Room's members
+    public List<Member> getMembers() {
+        return this.members;
+    }
+
+    //Get all of the members as a list of EmbeddedEntities
+    private List<EmbeddedEntity> getMembersAsEmbeddedEntities() {
+        return (List<EmbeddedEntity>) members.stream().map(Member::toEmbeddedEntity).collect(Collectors.toList());
+    }
+
+    //Returns the Room's video url list
+    public Queue<Video> getVideos() {
+        return this.videos;
+    }
+
+    //Returns the Room's message list
+    public LinkedList<Message> getMessages(){
+        return this.messages;
+    }
+
+    //Returns the messages as a list of embedded entities
+    private List<EmbeddedEntity> getMessagesAsEntities() {
+        return messages.stream().map(Message::toEmbeddedEntity).collect(Collectors.toList());
+    }
+
+    //Returns a queue of embedded entities
+    private List<EmbeddedEntity> getVideosAsEntities() {
+        return this.videos.stream().map(Video::toEmbeddedEntity).collect(Collectors.toList());
+    }
+
+    //Turns the Room object into a datastore entity
+    public static Entity toEntity(Room room){
+        Entity newRoom = new Entity(ROOM_ENTITY);
+        newRoom.setProperty(MEMBERS_PROPERTY, room.getMembersAsEmbeddedEntities());
+        newRoom.setProperty(VIDEOS_PROPERTY, room.getVideosAsEntities());
+        newRoom.setProperty(MESSAGES_PROPERTY, room.getMessagesAsEntities());
+        return newRoom;
+    }
+
+    //Creates a room object from a Datastore Key
+    public static Room fromKey(Key roomKey) {
+        try {
+            return Room.fromEntity(datastore.get(roomKey));
+        } catch (EntityNotFoundException e) {
+            System.out.println(e.toString());
+        }
+        return null;
+    }
+
+    // Helper addMessage function to manipulate the list of messages
+    private void addMessage(Message msg){
+        if(messages.size() >= 10) {
+            messages.remove(0);
```

Comment:
why not store message as a fixed capacity queue and pop earliest if needed?

Reply:
This could be done